# Python Pandas

- The *pandas* package is a important tool at the disposal of Data Scientists and Analysts working in Python today. 

- The powerful machine learning and glamorous visualization tools may get all the attention, but pandas is the backbone of most data projects. 

## What's Pandas for?

- Explore a dataset stored in a CSV. 

- Calculate statistics and answer questions about the data, like
    - What's the average, median, max, or min of each column? 
    - Does column A correlate with column B?
    - What does the distribution of data in column C look like?

- Clear the data by doing things like removing missing values and filtering rows or columns by some criteria.

- Visualize data with help of Matplotlib.

- Save the cleaned DataFrame, transformed data back into a CSV, other file or database.

## How does pandas fit into the data science toolkit?

Pandas library is used in conjunction with other Python libraries. 

Pandas is built on top of the **NumPy** package.

Data in pandas is often used to feed statistical analysis in **SciPy**, plotting functions from **Matplotlib**, and machine learning algorithms in **Scikit-learn**.

**Jupyter Notebooks** offer a good environment for using pandas to do data exploration and modeling, but pandas can also be used in text editors just as easily.

## Install and import

In [ ]:
# install
# pip install pandas

In [1]:
import pandas as pd

## Series and DataFrames

A `Series` is essentially a column.

A `DataFrame` is a multi-dimensional table made up of a collection of Series. 

### Creating DataFrames

There are *many* ways to create a DataFrame, but a great option is use a simple `dict`. 

Supposed that we have a fruit stand that sells apples and oranges. 
We want to have a column for each fruit and a row for each customer purchase. 

In [2]:
data = {
    'apples': [3, 2, 0, 1], 
    'oranges': [0, 3, 7, 2]
}

purchases = pd.DataFrame(data) # pandas DataFrame constructor

purchases

apples  oranges
0       3        0
1       2        3
2       0        7
3       1        2

Creating our own **index** when we initialize the DataFrame.

In [3]:
purchases = pd.DataFrame(data, index=['June', 'Robert', 'Lily', 'David'])

purchases

apples  oranges
June         3        0
Robert       2        3
Lily         0        7
David        1        2

We can **loc**ate a customer's order by using their name.

In [4]:
purchases.loc['June']

apples     3
oranges    0
Name: June, dtype: int64

### Save DataFrame in .csv format

In [5]:
# save DataFrame in .csv format
purchases.to_csv('purchases.csv')

## Reading data

### Reading data from CSVs

In [6]:
df = pd.read_csv('purchases.csv')

df

Unnamed: 0  apples  oranges
0       June       3        0
1     Robert       2        3
2       Lily       0        7
3      David       1        2

CSVs don't have indexes like DataFrames.

Use `index_col` to define indexes.

In [7]:
df = pd.read_csv('purchases.csv', index_col=0)

df

apples  oranges
June         3        0
Robert       2        3
Lily         0        7
David        1        2

## DataFrame operations

DataFrames possess hundreds of methods and other operations that are crucial to any analysis. 

Let's load in the IMDB movies dataset.

In [8]:
# load the IMDB movies dataset
movies_df = pd.read_csv("../data/IMDB-Movie-Data.csv", index_col="Title")

type(movies_df)

pandas.core.frame.DataFrame

### Viewing your data

In [9]:
# `.head()`: outputs the **first** five rows of your DataFrame by default
movies_df.head()

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   

                         Revenue (Millions)  Metascore  
Title                                                   
Guardians of the Galaxy              333.13       76.0  
Prometheus                           126.46       65.0  
Split                                138.12       62.0  
Sing                                 270.32       59.0  
Suicide Squad                        325.02       40.0

The `movies_df.head(10)` comand output the top ten rows. 

To see the **last** five rows use `movies_df.tail()`.

In [10]:
# .head(): outputs the **last** five rows of your DataFrame by default
movies_df.tail(2)

Rank                  Genre  \
Title                                       
Search Party   999       Adventure,Comedy   
Nine Lives    1000  Comedy,Family,Fantasy   

                                                    Description  \
Title                                                             
Search Party  A pair of friends embark on a mission to reuni...   
Nine Lives    A stuffy businessman finds himself trapped ins...   

                      Director  \
Title                            
Search Party    Scot Armstrong   
Nine Lives    Barry Sonnenfeld   

                                                         Actors  Year  \
Title                                                                   
Search Party  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
Nine Lives    Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

              Runtime (Minutes)  Rating  Votes  Revenue (Millions)  Metascore  
Title                                                                          
Search Party                 93     5.6   4881                 NaN       22.0  
Nine Lives                   87     5.3  12435               19.64       11.0

### Getting info about your data

`.info()`: return informations about the DataFrame.

In [11]:
movies_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, Guardians of the Galaxy to Nine Lives
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Genre               1000 non-null   object 
 2   Description         1000 non-null   object 
 3   Director            1000 non-null   object 
 4   Actors              1000 non-null   object 
 5   Year                1000 non-null   int64  
 6   Runtime (Minutes)   1000 non-null   int64  
 7   Rating              1000 non-null   float64
 8   Votes               1000 non-null   int64  
 9   Revenue (Millions)  872 non-null    float64
 10  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(4)
memory usage: 93.8+ KB


`.shape`: outputs a tuple of (rows, columns):

In [12]:
movies_df.shape

(1000, 11)

### Make DataFrame copy

`_append()`: return a copy without affecting the original DataFrame.

In [13]:
#temp_df = movies_df._append()

copie_movies_df = movies_df.copy()

temp_df = movies_df.copy()

temp_df.shape

(1000, 11)

### Handling duplicates

`drop_duplicates()` method: return a copy of your DataFrame with duplicates removed.

In [15]:
# dropping duplicates rows

temp1_df = temp_df.drop_duplicates()

temp1_df.shape

(1000, 11)

`inplace=True` argument: modify the DataFrame object in place.

In [16]:
temp_df.drop_duplicates(inplace=True)

### Column cleanup

Many times datasets will have verbose column names with symbols, upper and lowercase words, spaces, and typos.

In [17]:
# print the column names of our dataset
movies_df.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')

`.rename()` method: rename certain or all columns via a `dict`.

In [18]:
movies_df.rename(columns={
        'Runtime (Minutes)': 'Runtime', 
        'Revenue (Millions)': 'Revenue_millions'
    }, inplace=True)


movies_df.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year', 'Runtime',
       'Rating', 'Votes', 'Revenue_millions', 'Metascore'],
      dtype='object')

If we want to lowercase all names, instead of using `.rename()` we could also set a list of names to the columns or renaming each column manually we can do a list comprehension.

In [19]:
# renaming each column manually we can do a list comprehension
movies_df.columns = [col.lower() for col in movies_df]

movies_df.columns

Index(['rank', 'genre', 'description', 'director', 'actors', 'year', 'runtime',
       'rating', 'votes', 'revenue_millions', 'metascore'],
      dtype='object')

### Missing values

Most commonly missing values in DataFrames: `None` or `np.nan`.

There are two options in dealing with nulls: 

1. Get rid of rows or columns with nulls
2. Replace nulls with non-null values, a technique known as **imputation**

`isnull()` method: returns a DataFrame where each cell is either True or False depending on that cell's null status.

In [20]:
# to check which cells in our DataFrame are null
movies_df.isnull()

rank  genre  description  director  actors   year  \
Title                                                                         
Guardians of the Galaxy  False  False        False     False   False  False   
Prometheus               False  False        False     False   False  False   
Split                    False  False        False     False   False  False   
Sing                     False  False        False     False   False  False   
Suicide Squad            False  False        False     False   False  False   
...                        ...    ...          ...       ...     ...    ...   
Secret in Their Eyes     False  False        False     False   False  False   
Hostel: Part II          False  False        False     False   False  False   
Step Up 2: The Streets   False  False        False     False   False  False   
Search Party             False  False        False     False   False  False   
Nine Lives               False  False        False     False   False  False   

                         runtime  rating  votes  revenue_millions  metascore  
Title                                                                         
Guardians of the Galaxy    False   False  False             False      False  
Prometheus                 False   False  False             False      False  
Split                      False   False  False             False      False  
Sing                       False   False  False             False      False  
Suicide Squad              False   False  False             False      False  
...                          ...     ...    ...               ...        ...  
Secret in Their Eyes       False   False  False              True      False  
Hostel: Part II            False   False  False             False      False  
Step Up 2: The Streets     False   False  False             False      False  
Search Party               False   False  False              True      False  
Nine Lives                 False   False  False             False      False  

[1000 rows x 11 columns]

`.isnull().sum()` method: return the number of nulls in each column.

In [21]:
movies_df.isnull().sum()

rank                  0
genre                 0
description           0
director              0
actors                0
year                  0
runtime               0
rating                0
votes                 0
revenue_millions    128
metascore            64
dtype: int64

#### Removing null values

`.dropna() method`: return a new DataFrame without altering the original one where it delete any **row** with at least a single null value.

In [22]:
# remove 128 rows where `revenue_millions` is null and 64 rows where `metascore` is null.
drop1_movies_df = movies_df.dropna()
drop1_movies_df.shape

(838, 11)

In [23]:
movies_df.shape

(1000, 11)

The argument `axis=1` drop columns with null values.

In [24]:
# remove the `revenue_millions` and `metascore` columns.
drop2_movies_df = movies_df.dropna(axis=1)
drop2_movies_df.shape

(1000, 9)

If you specify `inplace=True` in this methods as well it return a new DataFrame altering the original.

### Imputation

Imputation is a conventional feature engineering technique used to keep valuable data that have null values. 

We can impute that null with another value, usually the **mean** or the **median** of that column. 

Let's look at imputing the missing values in the `revenue_millions` column. 

In [25]:
# Extracting a column into its own variable
revenue = movies_df['revenue_millions']

# a Series
revenue.head()

Title
Guardians of the Galaxy    333.13
Prometheus                 126.46
Split                      138.12
Sing                       270.32
Suicide Squad              325.02
Name: revenue_millions, dtype: float64

Impute the missing values of revenue using the mean. 

In [26]:
# calculating the mean
revenue_mean = revenue.mean()

revenue_mean

82.95637614678898

Let's fill the nulls using `fillna()`:

In [27]:
revenue.fillna(revenue_mean, inplace=True)

We have now replaced all nulls in `revenue` with the mean of the column. 

Using `inplace=True` we have actually affected the original `movies_df`.

In [28]:
movies_df.isnull().sum()

rank                 0
genre                0
description          0
director             0
actors               0
year                 0
runtime              0
rating               0
votes                0
revenue_millions     0
metascore           64
dtype: int64

### Understanding your variables

`describe()` method: return a summary of the distribution of continuous variables.

In [29]:
movies_df.describe()

rank         year      runtime       rating         votes  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1.000000e+03   
mean    500.500000  2012.783000   113.172000     6.723200  1.698083e+05   
std     288.819436     3.205962    18.810908     0.945429  1.887626e+05   
min       1.000000  2006.000000    66.000000     1.900000  6.100000e+01   
25%     250.750000  2010.000000   100.000000     6.200000  3.630900e+04   
50%     500.500000  2014.000000   111.000000     6.800000  1.107990e+05   
75%     750.250000  2016.000000   123.000000     7.400000  2.399098e+05   
max    1000.000000  2016.000000   191.000000     9.000000  1.791916e+06   

       revenue_millions   metascore  
count       1000.000000  936.000000  
mean          82.956376   58.985043  
std           96.412043   17.194757  
min            0.000000   11.000000  
25%           17.442500   47.000000  
50%           60.375000   59.500000  
75%           99.177500   72.000000  
max          936.630000  100.000000

In [30]:
#return the 
# count of rows, 
# unique count of categories, 
# top category, and 
# freq of top category. 
movies_df['genre'].describe()

count                        1000
unique                        207
top       Action,Adventure,Sci-Fi
freq                           50
Name: genre, dtype: object

The genre column has 207 unique values, the top value is Action/Adventure/Sci-Fi, which shows up 50 times (freq).

`.value_counts()` method: return the frequency of all values in a column.

In [31]:
movies_df['genre'].value_counts().head(10)

Action,Adventure,Sci-Fi       50
Drama                         48
Comedy,Drama,Romance          35
Comedy                        32
Drama,Romance                 31
Animation,Adventure,Comedy    27
Action,Adventure,Fantasy      27
Comedy,Drama                  27
Comedy,Romance                26
Crime,Drama,Thriller          24
Name: genre, dtype: int64

### DataFrame slicing, selecting, extracting

#### By column

To extract a column as a *Serie*, to pass a column names.

In [32]:
# return a Series
genre_col = movies_df['genre']

type(genre_col)

pandas.core.series.Series

 
To extract a column as a *DataFrame*, to pass a list of column names. 

In [33]:
genre_col = movies_df[['genre']]

type(genre_col)

pandas.core.frame.DataFrame

Since it's just a list, adding another column name is easy:

In [34]:
subset = movies_df[['genre', 'rating']]

subset.head()

genre  rating
Title                                                    
Guardians of the Galaxy   Action,Adventure,Sci-Fi     8.1
Prometheus               Adventure,Mystery,Sci-Fi     7.0
Split                             Horror,Thriller     7.3
Sing                      Animation,Comedy,Family     7.2
Suicide Squad            Action,Adventure,Fantasy     6.2

#### By rows

For rows, we have two options: 

- `.loc` - **loc**ates by name
- `.iloc`- **loc**ates by numerical **i**ndex

In [35]:
# `.loc`: use the Title of a movie
prom = movies_df.loc["Prometheus"]

prom

rank                                                                2
genre                                        Adventure,Mystery,Sci-Fi
description         Following clues to the origin of mankind, a te...
director                                                 Ridley Scott
actors              Noomi Rapace, Logan Marshall-Green, Michael Fa...
year                                                             2012
runtime                                                           124
rating                                                            7.0
votes                                                          485820
revenue_millions                                               126.46
metascore                                                        65.0
Name: Prometheus, dtype: object

In [36]:
# `iloc`: use the numerical index
prom = movies_df.iloc[1]
prom

rank                                                                2
genre                                        Adventure,Mystery,Sci-Fi
description         Following clues to the origin of mankind, a te...
director                                                 Ridley Scott
actors              Noomi Rapace, Logan Marshall-Green, Michael Fa...
year                                                             2012
runtime                                                           124
rating                                                            7.0
votes                                                          485820
revenue_millions                                               126.46
metascore                                                        65.0
Name: Prometheus, dtype: object

`loc` and `iloc` can be thought of as similar to Python `list` slicing. 

In [37]:
movie_subset = movies_df.loc['Prometheus':'Sing']
# movie_subset = movies_df.iloc[1:4]

movie_subset

rank                     genre  \
Title                                        
Prometheus     2  Adventure,Mystery,Sci-Fi   
Split          3           Horror,Thriller   
Sing           4   Animation,Comedy,Family   

                                                  description  \
Title                                                           
Prometheus  Following clues to the origin of mankind, a te...   
Split       Three girls are kidnapped by a man with a diag...   
Sing        In a city of humanoid animals, a hustling thea...   

                        director  \
Title                              
Prometheus          Ridley Scott   
Split         M. Night Shyamalan   
Sing        Christophe Lourdelet   

                                                       actors  year  runtime  \
Title                                                                          
Prometheus  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012      124   
Split       James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016      117   
Sing        Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016      108   

            rating   votes  revenue_millions  metascore  
Title                                                    
Prometheus     7.0  485820            126.46       65.0  
Split          7.3  157606            138.12       62.0  
Sing           7.2   60545            270.32       59.0

#### Conditional selections

Supposed that we want to filter on DataFrame all the films directed by Ridley Scott.

To do that, we take a column from the DataFrame and apply a Boolean condition to it.

In [ ]:
condition = (movies_df['director'] == "Ridley Scott")

condition.head()

In [ ]:
movies_df[movies_df['director'] == "Ridley Scott"].head()

Let's look at conditional selections using numerical values by filtering the DataFrame by ratings.

In [ ]:
movies_df[movies_df['rating'] >= 8.6].head(3)

We can make some conditionals by using logical operators `|` for "or" and `&` for "and".

Let's filter the the DataFrame to show only movies by Christopher Nolan OR Ridley Scott.

In [ ]:
movies_df[(movies_df['director'] == 'Christopher Nolan') | (movies_df['director'] == 'Ridley Scott')].head()

Using the `isin()` method we could make this more concise though.

In [ ]:
movies_df[movies_df['director'].isin(['Christopher Nolan', 'Ridley Scott'])].head()

## Applying functions

It is possible to iterate over a DataFrame or Series as you would with a list, but doing so is very slow.

An efficient alternative is to `apply()` a function to the dataset. 

For example, we could use a function to convert movies with an 8.0 or greater to a string value of "good" and the rest to "bad" and use this transformed values to create a new column.

In [ ]:
# creating a function that, when given a rating, determines if it's good or bad
def rating_function(x):
    if x >= 8.0:
        return "good"
    else:
        return "bad"

To send the entire rating column through this function.

In [ ]:
movies_df["rating_category"] = movies_df["rating"].apply(rating_function)

movies_df.head(2)

The `.apply()` method passes every value in the `rating` column through the `rating_function` and then returns a new Series. 
This Series is then assigned to a new column called `rating_category`.

## Plotting

Another great thing about pandas is that it integrates with Matplotlib, so you get the ability to plot directly off DataFrames and Series.

In [ ]:
import matplotlib.pyplot as plt

# set font and plot size to be larger
plt.rcParams.update({'font.size': 20, 'figure.figsize': (12, 8)}) 

Plot a Scatterplots based on two columns.

In [ ]:
movies_df.plot(
    kind='scatter', 
    x='rating', 
    y='revenue_millions', 
    title='Revenue (millions) vs Rating'
    );

Plot a Histogram based on a single column.

In [ ]:
movies_df['rating'].plot(kind='hist', title='Rating');

In [ ]:
from scipy.integrate import quad
def integrand(x, a, b):
    return a*x**2 + b

## Reference

https://pandas.pydata.org/docs/index.html

https://www.learndatasci.com/tutorials/python-pandas-tutorial-complete-introduction-for-beginners/

https://pandas.pydata.org/docs/user_guide/10min.html

https://pandas.pydata.org/docs/getting_started/intro_tutorials/index.html
